File: backends/gin/internal/service/plaza.go
package service

import (
	"strings"
	"time"

	"gorm.io/gorm"

	"hellotime/gin/internal/core"
	"hellotime/gin/internal/dto"
	"hellotime/gin/internal/model"
)

// PlazaRow 是 JOIN 查询返回的单行（capsule + owner 信息）。
type PlazaRow struct {
	model.Capsule
	OwnerNickname string
	OwnerAvatarID string
}

func buildPlazaQuery(db *gorm.DB) *gorm.DB {
	return db.Table("capsules").
		Select("capsules.*, users.nickname as owner_nickname, users.avatar_id as owner_avatar_id").
		Joins("JOIN users ON users.id = capsules.owner_id").
		Where("capsules.in_plaza = ?", true)
}

func applyFilter(q *gorm.DB, filter string) (*gorm.DB, error) {
	now := time.Now().UTC()
	switch filter {
	case "opened":
		return q.Where("capsules.open_at <= ?", now), nil
	case "unopened":
		return q.Where("capsules.open_at > ?", now), nil
	case "all", "":
		return q, nil
	default:
		return nil, core.ValidationErr("filter 仅支持 all/opened/unopened", "filter")
	}
}

func applySearch(q *gorm.DB, search string) (*gorm.DB, error) {
	s := strings.TrimSpace(search)
	if s == "" {
		return q, nil
	}
	if len([]rune(s)) > 50 {
		return nil, core.ValidationErr("q 长度不得超过 50", "q")
	}
	pattern := "%" + strings.ToLower(s) + "%"
	return q.Where("LOWER(capsules.title) LIKE ? OR LOWER(users.nickname) LIKE ?", pattern, pattern), nil
}

// PlazaList 广场列表（分页 + 排序 + 筛选 + 搜索）。
func PlazaList(db *gorm.DB, sort, filter, q string, page, pageSize int, viewerID string) (*dto.Paginated[dto.CapsuleListItem], error) {
	if sort != "hot" && sort != "new" {
		return nil, core.ValidationErr("sort 仅支持 hot/new", "sort")
	}
	if page < 1 {
		return nil, core.ValidationErr("page 必须 >= 1", "page")
	}
	if pageSize < 1 || pageSize > 50 {
		return nil, core.ValidationErr("pageSize 范围 1-50", "pageSize")
	}

	base := buildPlazaQuery(db)
	var err error
	if base, err = applyFilter(base, filter); err != nil {
		return nil, err
	}
	if base, err = applySearch(base, q); err != nil {
		return nil, err
	}

	var total int64
	if err := base.Count(&total).Error; err != nil {
		return nil, core.InternalErr("查询失败")
	}

	if sort == "hot" {
		base = base.Order("capsules.favorite_count DESC, capsules.created_at DESC")
	} else {
		base = base.Order("capsules.created_at DESC")
	}

	var rows []PlazaRow
	if err := base.Offset((page - 1) * pageSize).Limit(pageSize).Scan(&rows).Error; err != nil {
		return nil, core.InternalErr("查询失败")
	}

	capsuleIDs := make([]string, len(rows))
	for i, r := range rows {
		capsuleIDs[i] = r.ID
	}
	favMap := isFavoritedByViewer(db, viewerID, capsuleIDs)

	items := make([]dto.CapsuleListItem, len(rows))
	for i, r := range rows {
		owner := model.User{Nickname: r.OwnerNickname, AvatarID: r.OwnerAvatarID}
		items[i] = capsuleToListItem(&r.Capsule, &owner, favMap[r.ID], nil)
	}

	return &dto.Paginated[dto.CapsuleListItem]{
		Items: items,
		Pagination: dto.Pagination{
			Page:       page,
			PageSize:   pageSize,
			Total:      int(total),
			TotalPages: dto.CalcTotalPages(int(total), pageSize),
		},
	}, nil
}

// MyCapsules 返回当前用户创建的胶囊列表（分页）。
func MyCapsules(db *gorm.DB, user *model.User, page, pageSize int) (*dto.Paginated[dto.CapsuleListItem], error) {
	if page < 1 {
		return nil, core.ValidationErr("page 必须 >= 1", "page")
	}
	if pageSize < 1 || pageSize > 50 {
		return nil, core.ValidationErr("pageSize 范围 1-50", "pageSize")
	}

	var total int64
	db.Model(&model.Capsule{}).Where("owner_id = ?", user.ID).Count(&total)

	var rows []PlazaRow
	db.Table("capsules").
		Select("capsules.*, users.nickname as owner_nickname, users.avatar_id as owner_avatar_id").
		Joins("JOIN users ON users.id = capsules.owner_id").
		Where("capsules.owner_id = ?", user.ID).
		Order("capsules.created_at DESC").
		Offset((page - 1) * pageSize).Limit(pageSize).
		Scan(&rows)

	items := make([]dto.CapsuleListItem, len(rows))
	for i, r := range rows {
		owner := model.User{Nickname: r.OwnerNickname, AvatarID: r.OwnerAvatarID}
		items[i] = capsuleToListItem(&r.Capsule, &owner, false, nil)
	}

	return &dto.Paginated[dto.CapsuleListItem]{
		Items: items,
		Pagination: dto.Pagination{
			Page:       page,
			PageSize:   pageSize,
			Total:      int(total),
			TotalPages: dto.CalcTotalPages(int(total), pageSize),
		},
	}, nil
}

// FavoriteRow 是收藏 JOIN 查询行。
type FavoriteRow struct {
	model.Capsule
	OwnerNickname string
	OwnerAvatarID string
	FavCreatedAt  time.Time
}

// MyFavorites 返回当前用户收藏的胶囊列表（分页）。
func MyFavorites(db *gorm.DB, user *model.User, page, pageSize int) (*dto.Paginated[dto.CapsuleListItem], error) {
	if page < 1 {
		return nil, core.ValidationErr("page 必须 >= 1", "page")
	}
	if pageSize < 1 || pageSize > 50 {
		return nil, core.ValidationErr("pageSize 范围 1-50", "pageSize")
	}

	var total int64
	db.Model(&model.Favorite{}).Where("user_id = ?", user.ID).Count(&total)

	var rows []FavoriteRow
	db.Table("favorites").
		Select("capsules.*, users.nickname as owner_nickname, users.avatar_id as owner_avatar_id, favorites.created_at as fav_created_at").
		Joins("JOIN capsules ON capsules.id = favorites.capsule_id").
		Joins("JOIN users ON users.id = capsules.owner_id").
		Where("favorites.user_id = ?", user.ID).
		Order("favorites.created_at DESC").
		Offset((page - 1) * pageSize).Limit(pageSize).
		Scan(&rows)

	items := make([]dto.CapsuleListItem, len(rows))
	for i, r := range rows {
		owner := model.User{Nickname: r.OwnerNickname, AvatarID: r.OwnerAvatarID}
		favAt := r.FavCreatedAt
		items[i] = capsuleToListItem(&r.Capsule, &owner, true, &favAt)
	}

	return &dto.Paginated[dto.CapsuleListItem]{
		Items: items,
		Pagination: dto.Pagination{
			Page:       page,
			PageSize:   pageSize,
			Total:      int(total),
			TotalPages: dto.CalcTotalPages(int(total), pageSize),
		},
	}, nil
}

File: backends/elysia/src/services/plaza.ts
import { one, query } from "../db";
import { ERR } from "../errors";
import { capsuleListItem, CapsuleRow, iso, OwnerBrief, pagination } from "../types";
import { rowOwner } from "./capsules";

function validatePage(page: number, pageSize: number) {
  if (!Number.isInteger(page) || page < 1) throw ERR.validation("page 必须 >= 1", "page");
  if (!Number.isInteger(pageSize) || pageSize < 1 || pageSize > 50) {
    throw ERR.validation("pageSize 范围 1-50", "pageSize");
  }
}

async function favoriteSet(viewerId: string | null, ids: string[]) {
  if (!viewerId || ids.length === 0) return new Set<string>();
  const marks = ids.map(() => "?").join(",");
  const rows = await query<{ capsuleId: string }>(
    `SELECT capsule_id AS "capsuleId" FROM favorites
      WHERE user_id = ? AND capsule_id IN (${marks})`,
    [viewerId, ...ids],
  );
  return new Set(rows.map((r) => r.capsuleId));
}

export async function plazaList(opts: {
  sort: string;
  filter: string;
  q: string | null;
  page: number;
  pageSize: number;
  viewerId: string | null;
}) {
  if (!["hot", "new"].includes(opts.sort)) throw ERR.validation("sort 仅支持 hot/new", "sort");
  if (!["all", "opened", "unopened"].includes(opts.filter)) {
    throw ERR.validation("filter 仅支持 all/opened/unopened", "filter");
  }
  validatePage(opts.page, opts.pageSize);
  const q = (opts.q ?? "").trim();
  if (q.length > 50) throw ERR.validation("q 长度不得超过 50", "q");

  const where: string[] = ["c.in_plaza = ?"];
  const params: unknown[] = [true];
  const now = new Date().toISOString();
  if (opts.filter === "opened") {
    where.push("c.open_at <= ?");
    params.push(now);
  } else if (opts.filter === "unopened") {
    where.push("c.open_at > ?");
    params.push(now);
  }
  if (q) {
    where.push("(lower(c.title) LIKE ? OR lower(u.nickname) LIKE ?)");
    params.push(`%${q.toLowerCase()}%`, `%${q.toLowerCase()}%`);
  }
  const clause = where.join(" AND ");
  const order =
    opts.sort === "hot"
      ? "c.favorite_count DESC, c.created_at DESC"
      : "c.created_at DESC";
  const rows = await query<CapsuleRow & OwnerBrief>(
    `SELECT c.id, c.owner_id AS "ownerId", c.code, c.title, c.content,
            c.open_at AS "openAt", c.in_plaza AS "inPlaza",
            c.favorite_count AS "favoriteCount", c.created_at AS "createdAt",
            c.updated_at AS "updatedAt",
            u.nickname, u.avatar_id AS "avatarId"
       FROM capsules c
       JOIN users u ON u.id = c.owner_id
      WHERE ${clause}
      ORDER BY ${order}
      LIMIT ? OFFSET ?`,
    [...params, opts.pageSize, (opts.page - 1) * opts.pageSize],
  );
  const count = await one<{ n: string | number }>(
    `SELECT count(*) AS n FROM capsules c JOIN users u ON u.id = c.owner_id WHERE ${clause}`,
    params,
  );
  const faved = await favoriteSet(opts.viewerId, rows.map((r) => r.id));
  return {
    items: rows.map((r) => capsuleListItem(r, rowOwner(r), faved.has(r.id))),
    pagination: pagination(Number(count?.n ?? 0), opts.page, opts.pageSize),
  };
}

export async function myCapsules(userId: string, page: number, pageSize: number) {
  validatePage(page, pageSize);
  const count = await one<{ n: string | number }>(`SELECT count(*) AS n FROM capsules WHERE owner_id = ?`, [
    userId,
  ]);
  const rows = await query<CapsuleRow & OwnerBrief>(
    `SELECT c.id, c.owner_id AS "ownerId", c.code, c.title, c.content,
            c.open_at AS "openAt", c.in_plaza AS "inPlaza",
            c.favorite_count AS "favoriteCount", c.created_at AS "createdAt",
            c.updated_at AS "updatedAt",
            u.nickname, u.avatar_id AS "avatarId"
       FROM capsules c
       JOIN users u ON u.id = c.owner_id
      WHERE c.owner_id = ?
      ORDER BY c.created_at DESC
      LIMIT ? OFFSET ?`,
    [userId, pageSize, (page - 1) * pageSize],
  );
  return {
    items: rows.map((r) => capsuleListItem(r, rowOwner(r), false)),
    pagination: pagination(Number(count?.n ?? 0), page, pageSize),
  };
}

export async function myFavorites(userId: string, page: number, pageSize: number) {
  validatePage(page, pageSize);
  const count = await one<{ n: string | number }>(`SELECT count(*) AS n FROM favorites WHERE user_id = ?`, [
    userId,
  ]);
  const rows = await query<CapsuleRow & OwnerBrief & { favoritedAt: string }>(
    `SELECT c.id, c.owner_id AS "ownerId", c.code, c.title, c.content,
            c.open_at AS "openAt", c.in_plaza AS "inPlaza",
            c.favorite_count AS "favoriteCount", c.created_at AS "createdAt",
            c.updated_at AS "updatedAt",
            u.nickname, u.avatar_id AS "avatarId",
            f.created_at AS "favoritedAt"
       FROM favorites f
       JOIN capsules c ON c.id = f.capsule_id
       JOIN users u ON u.id = c.owner_id
      WHERE f.user_id = ?
      ORDER BY f.created_at DESC
      LIMIT ? OFFSET ?`,
    [userId, pageSize, (page - 1) * pageSize],
  );
  return {
    items: rows.map((r) => capsuleListItem(r, rowOwner(r), true, iso(r.favoritedAt))),
    pagination: pagination(Number(count?.n ?? 0), page, pageSize),
  };
}
